### PR TITLE
rec: drop ref in mtasker when it is no longer needed

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

If we do not do that, only the next call of `sendEvent` will release
the shared pointer to the `TCPConnectionHandler` in `PacketID` as it overwrites `d_eventkey`. Same for `d_waitval`. This
might delay cleaning TCP objects up on not-so-busy threads.

Fixes https://github.com/PowerDNS/pdns/issues/13422, thought it can still take a while on idle recursors,
as the housekeeping function is not called often in those cases.

Added a simple mechanism to make sure `d_eventkey` is not used when it is not set yet or after being moved.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/mtasker.hh [hunk 1]
```diff
@@ -332,18 +334,18 @@ void MTasker<Key, Val, Cmp>::yield()
...
   d_waitstatus = Answer;
   if (val) {
     d_waitval = *val;
   }
   d_tid = waiter->tid; // set tid
-  d_eventkey = waiter->key; // pass waitEvent the exact key it was woken for
+  d_eventkey = std::move(waiter->key); // pass waitEvent the exact key it was woken for
+  d_used = false;
   auto userspace = std::move(waiter->context);
   d_waiters.erase(waiter); // removes the waitpoint
```
Comment (rgacogne) [inline pdns/recursordist/mtasker.hh:350]: Calling `erase()` after the key has been moved worries me. In theory it should be OK given that we use an iterator, but it feels brittle. Do we really need to move the key here, given that the entry is going to be erased anyway?

File: pdns/recursordist/mtasker.hh [hunk 2]
```diff
@@ -457,14 +459,13 @@ bool MTasker<Key, Val, Cmp>::schedule(const struct timeval& now)
...
-    for (typename waiters_by_ttd_index_t::iterator i = ttdindex.begin(); i != ttdindex.end();) {
+    for (auto i = ttdindex.begin(); i != ttdindex.end();) {
       if (i->ttd.tv_sec && i->ttd < now) {
         d_waitstatus = TimeOut;
-        d_eventkey = i->key; // pass waitEvent the exact key it was woken for
+        d_eventkey = std::move(i->key); // pass waitEvent the exact key it was woken for
+        d_used = false;
         auto ucontext = i->context;
         d_tid = i->tid;
         ttdindex.erase(i++); // removes the waitpoint
```
Comment (rgacogne) [inline pdns/recursordist/mtasker.hh:471]: Same thing here.

File: pdns/recursordist/mtasker.hh (hunk 1 above)
Comment (omoerbeek) [reply pdns/recursordist/mtasker.hh:350]: On second thought, I agree.

File: pdns/recursordist/mtasker.hh (hunk 2 above)
Comment (omoerbeek) [reply pdns/recursordist/mtasker.hh:471]: Same

**Review by rgacogne — APPROVED** [review]
Looks good to me!
If it's not a lot of work (I have no idea how hard it would be), adding a unit test to ensure that we do not re-introduce the same issue in the future might be a good idea.

**Review by rgacogne — APPROVED** [review]

**Pull request merged**